pilot/pkg/model/push_context.go

```diff
@@ -78,9 +80,127 @@ type PushContext struct {
+
+	// WorkloadUpdate is called by a registry when the labels or annotations on a workload have changed.
+	// The 'id' is the IP address of the pod for k8s if the pod is in the main/default network.
+	// In future it will include the 'network id' for pods in a different network, behind a zvpn gate.
+	// The IP is used because K8S Endpoints object associated with a Service only include the IP.
+	// We use Endpoints to track the membership to a service and readiness.
+	WorkloadUpdate(id string, labels map[string]string, annotations map[string]string)
+}
+
+// IstioEndpoint has the information about a single address+port for a specific
+// service and shard. Unlike ServiceInstance, it has less info and require less
+// allocation.
+type IstioEndpoint struct {
+
+	// Labels points to the workload or deployment labels.
+	Labels map[string]string
+
+	// Family indicates what type of endpoint, such as TCP or Unix Domain Socket.
+	// Default is TCP.
+	Family AddressFamily
+
+	// Address is the address of the endpoint, using envoy proto.
+	Address string
+
+	// EndpointPort is the port where the workload is listening, can be different
+	// from the service port.
+	EndpointPort uint32
+
+	// ServicePortName tracks the name of the port, to avoid 'eventual consistency' issues.
+	// Sometimes the Endpoint is visible before Service - so looking up the port number would
+	// fail. Instead the mapping to number is made when the clusters are computed. The lazy
+	// computation will also help with 'on-demand' and 'split horizon' - where it will be skipped
+	// for not used clusters or endpoints behind a gate.
+	ServicePortName string
+
+	// UID identifies the workload, for telemetry purpose.
+	UID string
+
+	// EnvoyEndpoint is a cached LbEndpoint, converted from the data, to
+	// avoid recomputation
+	EnvoyEndpoint *endpoint.LbEndpoint
+
+	// ServiceAccount holds the associated service account.
+	// May be replaced once a better optimization/struct is available for secure naming.
+	ServiceAccount string
+}
+
+// ServiceShards holds the set of endpoint shards of a service. Registries update
+// individual shards incrementally. The shards are aggregated and split into
+// clusters when a push for the specific cluster is needed.
+type ServiceShards struct {
+
+	// Shards is used to track the shards. EDS updates are grouped by shard.
+	// Current implementation uses the registry name as key - in multicluster this is the
+	// name of the k8s cluster, derived from the config (secret).
+	Shards map[string]*EndpointShard
+
+	// ServiceAccounts has the concatenation of all service accounts seen so far in endpoints.
+	// This is updated on push, based on shards. If the previous list is different than
+	// current list, a full push will be forced, to trigger a secure naming update.
+	// Due to the larger time, it is still possible that connection errors will occur while
+	// CDS is updated.
+	ServiceAccounts map[string]bool
```

Comment:
store the name of the service in this struct, to avoid looking it up somewehre else or passing it as another variable in each function that uses this.